odoo/addons/test_read_group/tests/test_web_fill_temporal.py

```diff
@@ -0,0 +1,920 @@
+        """
+        self.Model.create({'datetime': '2017-12-31 21:00:00', 'value': 42})
+
+        expected = [{
+            '__extra_domain': ['&',
+                         ('datetime', '>=', '2017-12-31 17:00:00'),
+                         ('datetime', '<', '2018-01-31 17:00:00')],
+            'datetime:month': ('2017-12-31 17:00:00', 'January 2018'),
+            '__count': 1,
+            'value:sum': 42,
+        }]
+
+        model_fill = self.Model.with_context(tz='Asia/Hovd', fill_temporal=True)
+        groups = model_fill.web_read_group([], ['datetime:month'], ['__count', 'value:sum'])['groups']
+        self.assertEqual(groups, expected)
+
+    def test_with_bounds(self):
+        """Test the alternative dictionary format for the fill_temporal context key (fill_from, fill_to).
+
+        We apply the fill_temporal logic only to a cibled portion of the result of a web_read_group.
+        [fill_from, fill_to] are the inclusive bounds of this portion.
+        Data outside those bounds will not be filtered out
+        Bounds will be converted to the start of the period which they belong to (depending
+        on the granularity of the groupby). This means that we can put any date of the period as the bound
+        and it will still work.
+        """
+        self.Model.create({'date': '1916-02-15', 'value': 1})
+        self.Model.create({'date': '1916-06-15', 'value': 2})
+        self.Model.create({'date': '1916-11-15', 'value': 3})
+
+        expected = [{
+            '__extra_domain': ['&', ('date', '>=', '1916-02-01'), ('date', '<', '1916-03-01')],
+            'date:month': ('1916-02-01', 'February 1916'),
+            '__count': 1,
+            'value:sum': 1,
+        }, {
+            '__extra_domain': ['&', ('date', '>=', '1916-05-01'), ('date', '<', '1916-06-01')],
+            'date:month': ('1916-05-01', 'May 1916'),
+            '__count': 0,
+            'value:sum': False,
+        }, {
+            '__extra_domain': ['&', ('date', '>=', '1916-06-01'), ('date', '<', '1916-07-01')],
+            'date:month': ('1916-06-01', 'June 1916'),
+            '__count': 1,
+            'value:sum': 2,
+        }, {
+            '__extra_domain': ['&', ('date', '>=', '1916-07-01'), ('date', '<', '1916-08-01')],
+            'date:month': ('1916-07-01', 'July 1916'),
+            '__count': 0,
+            'value:sum': False,
+        }, {
+            '__extra_domain': ['&', ('date', '>=', '1916-08-01'), ('date', '<', '1916-09-01')],
+            'date:month': ('1916-08-01', 'August 1916'),
+            '__count': 0,
+            'value:sum': False,
+        }, {
+            '__extra_domain': ['&', ('date', '>=', '1916-11-01'), ('date', '<', '1916-12-01')],
+            'date:month': ('1916-11-01', 'November 1916'),
+            '__count': 1,
+            'value:sum': 3,
+        }]
+
+        groups = self.Model.with_context(fill_temporal={"fill_from": '1916-05-15', "fill_to": '1916-08-15'}).web_read_group(
```

Comment:
Missing test case where fill_temporal contains datetime values.

Reply:
Well, it doesn't make sense since fill_temporal should be only used from RPC.